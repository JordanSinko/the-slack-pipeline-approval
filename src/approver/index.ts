import qs from "querystring";
import CodePipeline from "aws-sdk/clients/codepipeline";
import phin from "phin";
import { APIGatewayEvent } from "aws-lambda";

const { AWS_REGION, AWS_ACCOUNT_ID } = process.env;
const { SLACK_TOKEN } = process.env;

enum Status {
  Approved = "Approved",
  Rejected = "Rejected",
}

enum StatusColor {
  Approved = "#007a5a",
  Rejected = "#9e1e1e",
}

const codepipeline = new CodePipeline({ region: AWS_REGION });

export const handler = async (event: APIGatewayEvent) => {
  if (event.body != null) {
    const body = event.isBase64Encoded === true ? Buffer.from(event.body, "base64").toString("utf8") : event.body;
    const { payload } = qs.parse(body) as { payload: string };
    const parsed = JSON.parse(payload);

    const { actions, user, channel, response_url } = parsed;
    const [action] = actions;
    const { approved, context } = JSON.parse(action.value);
    const { stageName, actionName, pipelineName, token, consoleLink, externalEntityLink, userInformation } = context;

    const status = approved ? Status.Approved : Status.Rejected;
    const statusColor = approved ? StatusColor.Approved : StatusColor.Rejected;

    let statusMessage = approved ? ":tallyhappy: Approved" : ":tallysad: Rejected";

    if (channel?.name !== "directmessage") {
      statusMessage += ` by @${user.name}`;
    }

    await codepipeline
      .putApprovalResult({
        pipelineName,
        stageName,
        actionName,
        token,
        result: {
          status,
          summary: statusMessage,
        },
      })
      .promise();

    await phin({
      method: "POST",
      url: response_url,
      headers: { "Content-Type": "application/json", Authorization: `Bearer ${SLACK_TOKEN}` },
      parse: "json",
      data: {
        replace_original: true,
        attachments: [
          {
            color: statusColor,
            blocks: [
              {
                type: "section",
                text: {
                  type: "mrkdwn",
                  text: `*<${consoleLink} | CodePipeline Approval | ${pipelineName} | Region: ${AWS_REGION} | Account: ${AWS_ACCOUNT_ID}>*`,
                },
              },
              {
                type: "section",
                fields: [
                  {
                    type: "mrkdwn",
                    text: `*Pipeline:*\n${pipelineName}`,
                  },
                  {
                    type: "mrkdwn",
                    text: `*Stage:*\n${stageName}`,
                  },
                  {
                    type: "mrkdwn",
                    text: `*Action:*\n${actionName}`,
                  },
                ],
              },
              userInformation && {
                type: "section",
                fields: [
                  {
                    type: "mrkdwn",
                    text: `*Additional information*\n${userInformation}`,
                  },
                ],
              },
              externalEntityLink && {
                type: "section",
                fields: [
                  {
                    type: "mrkdwn",
                    text: `*Content to review*\n${externalEntityLink}`,
                  },
                ],
              },
              {
                type: "section",
                fields: [
                  {
                    type: "mrkdwn",
                    text: `*Status*\n${statusMessage}`,
                  },
                ],
              },
            ].filter(Boolean),
          },
        ],
      } as Record<string, any>,
    });
  }
};
