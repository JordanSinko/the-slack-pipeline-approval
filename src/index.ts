import { resolve } from "path";
import { Stack, Construct } from "@aws-cdk/core";
import { IStage, ActionBindOptions, ActionConfig } from "@aws-cdk/aws-codepipeline";
import { ManualApprovalAction, ManualApprovalActionProps } from "@aws-cdk/aws-codepipeline-actions";
import { Topic } from "@aws-cdk/aws-sns";
import { Code, Function as AwsFunction, Runtime } from "@aws-cdk/aws-lambda";
import { HttpApi, HttpMethod, LambdaProxyIntegration } from "@aws-cdk/aws-apigatewayv2";
import { LambdaSubscription } from "@aws-cdk/aws-sns-subscriptions";
import { PolicyStatement } from "@aws-cdk/aws-iam";

export interface SlackApprovalActionProps extends ManualApprovalActionProps {
  readonly slackChannel: string;
  readonly slackUsername?: string;
  readonly slackIcon?: string;
}

export class SlackApprovalAction extends ManualApprovalAction {
  slackProps: SlackApprovalActionProps;
  approval: SlackApproval;
  id: string;
  slackAdditionalInformation?: string;

  constructor(approval: SlackApproval, id: string, props: SlackApprovalActionProps) {
    const { slackChannel, slackUsername, slackIcon, additionalInformation: _additionalInformation, ..._props } = props;
    const additionalInformation = JSON.stringify({
      slackChannel,
      slackUsername,
      slackIcon,
      userInformation: _additionalInformation,
    });

    super({
      ..._props,
      additionalInformation,
    });

    this.slackProps = props;
    this.id = id;
    this.approval = approval;
    this.slackAdditionalInformation = additionalInformation;
  }

  protected bound(_scope: Construct, stage: IStage, options: ActionBindOptions): ActionConfig {
    this.approval.topic.grantPublish(options.role);
    this.approval.approverHandler.role?.addToPrincipalPolicy(
      new PolicyStatement({
        actions: ["codepipeline:PutApprovalResult"],
        resources: [`${stage.pipeline.pipelineArn}/${stage.stageName}/${this.slackProps.actionName}`],
      })
    );

    return {
      configuration: undefinedIfAllValuesAreEmpty({
        NotificationArn: this.approval.topic.topicArn,
        CustomData: this.slackAdditionalInformation,
        ExternalEntityLink: this.slackProps.externalEntityLink,
      }),
    };
  }
}

export interface SlackApprovalProps {
  readonly slackToken: string;
}

export class SlackApproval extends Construct {
  public readonly props: SlackApprovalProps;
  public readonly topic: Topic;
  public readonly approverHandler: AwsFunction;

  constructor(scope: Construct, id: string, props: SlackApprovalProps) {
    super(scope, id);

    this.props = props;

    const stack = Stack.of(scope);
    const topic = new Topic(scope, `Topic`, {});

    this.topic = topic;

    const requester = new AwsFunction(scope, `Requester`, {
      runtime: Runtime.NODEJS_12_X,
      code: Code.fromAsset(resolve(__dirname, "requester"), {}),
      handler: "index.handler",
      environment: {
        AWS_ACCOUNT_ID: stack.account,
        SLACK_TOKEN: this.props.slackToken,
      },
    });

    topic.addSubscription(new LambdaSubscription(requester));

    const api = new HttpApi(this, "Api", {});

    const approver = new AwsFunction(this, "Approver", {
      runtime: Runtime.NODEJS_12_X,
      code: Code.fromAsset(resolve(__dirname, "approver"), {}),
      handler: "index.handler",
      environment: {
        AWS_ACCOUNT_ID: stack.account,
        SLACK_TOKEN: this.props.slackToken,
      },
    });

    this.approverHandler = approver;

    api.addRoutes({
      path: "/{proxy+}",
      methods: [HttpMethod.ANY],
      integration: new LambdaProxyIntegration({ handler: approver }),
    });
  }

  public addApprovalAction(id: string, props: SlackApprovalActionProps) {
    const action = new SlackApprovalAction(this, id, { ...props });
    return action;
  }
}

function undefinedIfAllValuesAreEmpty(object: object): object | undefined {
  return Object.values(object).some((v) => v !== undefined) ? object : undefined;
}
