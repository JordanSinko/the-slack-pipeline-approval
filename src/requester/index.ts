import phin from "phin";
import { SNSEvent } from "aws-lambda";

const { AWS_REGION, AWS_ACCOUNT_ID } = process.env;
const { SLACK_TOKEN } = process.env;

interface Approval {
  consoleLink: string;
  pipelineName: string;
  stageName: string;
  actionName: string;
  expires: string;
  externalEntityLink?: string;
  customData?: string;
}

interface ApprovalCustomData {
  slackChannel: string;
  slackUsername?: string;
  slackIcon?: string;
  userInformation?: string;
}

export const handler = async (event: SNSEvent) => {
  const message = event.Records[0].Sns.Message;
  const { consoleLink, approval } = JSON.parse(message) as { consoleLink: string; approval: Approval };
  const {
    pipelineName: pipeline,
    stageName: stage,
    actionName: action,
    expires,
    externalEntityLink,
    customData,
  } = approval;

  const approvalContext = { ...approval, consoleLink, region: AWS_REGION, account: AWS_ACCOUNT_ID };
  const data: ApprovalCustomData = JSON.parse(customData ?? "{}");

  const { slackChannel, slackIcon, slackUsername, userInformation } = data;

  await phin({
    method: "POST",
    url: "https://slack.com/api/chat.postMessage",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${SLACK_TOKEN}` },
    parse: "json",
    data: {
      channel: slackChannel,
      username: slackUsername,
      icon_emoji: slackIcon,
      attachments: [
        {
          blocks: [
            {
              type: "section",
              text: {
                type: "mrkdwn",
                text: `*<${consoleLink} | CodePipeline Approval | ${pipeline} | Region: ${AWS_REGION} | Account: ${AWS_ACCOUNT_ID}>*`,
              },
            },
            {
              type: "section",
              fields: [
                {
                  type: "mrkdwn",
                  text: `*Pipeline:*\n${pipeline}`,
                },
                {
                  type: "mrkdwn",
                  text: `*Stage:*\n${stage}`,
                },
                {
                  type: "mrkdwn",
                  text: `*Action:*\n${action}`,
                },
                {
                  type: "mrkdwn",
                  text: `*Expires:*\n${expires}`,
                },
              ],
            },
            userInformation && {
              type: "section",
              fields: [
                {
                  type: "mrkdwn",
                  text: `*Additional information*\n${userInformation}`,
                },
              ],
            },
            externalEntityLink && {
              type: "section",
              fields: [
                {
                  type: "mrkdwn",
                  text: `*Content to review*\n${externalEntityLink}`,
                },
              ],
            },
            {
              type: "section",
              fields: [
                {
                  type: "mrkdwn",
                  text: `*Status*\n:hourglass: Pending`,
                },
              ],
            },
            {
              type: "actions",
              elements: [
                {
                  type: "button",
                  text: {
                    type: "plain_text",
                    emoji: true,
                    text: "Approve",
                  },
                  style: "primary",
                  value: `${JSON.stringify({ approved: true, context: approvalContext })}`,
                },
                {
                  type: "button",
                  text: {
                    type: "plain_text",
                    emoji: true,
                    text: "Deny",
                  },
                  style: "danger",
                  value: `${JSON.stringify({ approved: false, context: approvalContext })}`,
                },
              ],
            },
          ].filter(Boolean),
        },
      ],
    } as Record<string, any>,
  });

  return;
};
