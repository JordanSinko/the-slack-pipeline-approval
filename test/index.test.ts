import "@aws-cdk/assert/jest";
import { App, Stack } from "@aws-cdk/core";

import { SlackApproval } from "../src";

test("Snapshot", () => {
  const app = new App();
  const stack = new Stack(app, "testing-stack");

  const approval = new SlackApproval(stack, "Approval", { slackToken: "xoxb-12345" });
  approval.addApprovalAction("Approval", { actionName: "Approval", slackChannel: "1" });

  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});
