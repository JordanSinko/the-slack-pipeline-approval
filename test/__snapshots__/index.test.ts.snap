// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters9df69b1c5244ca159360ab80f7e94d352de6fc13390ac375c21274ed8bed348dArtifactHashEEC8ED0C": Object {
      "Description": "Artifact hash for asset \\"9df69b1c5244ca159360ab80f7e94d352de6fc13390ac375c21274ed8bed348d\\"",
      "Type": "String",
    },
    "AssetParameters9df69b1c5244ca159360ab80f7e94d352de6fc13390ac375c21274ed8bed348dS3Bucket01FB5559": Object {
      "Description": "S3 bucket for asset \\"9df69b1c5244ca159360ab80f7e94d352de6fc13390ac375c21274ed8bed348d\\"",
      "Type": "String",
    },
    "AssetParameters9df69b1c5244ca159360ab80f7e94d352de6fc13390ac375c21274ed8bed348dS3VersionKey4C30C2E2": Object {
      "Description": "S3 key for asset version \\"9df69b1c5244ca159360ab80f7e94d352de6fc13390ac375c21274ed8bed348d\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ApprovalApi7AD56316": Object {
      "Properties": Object {
        "Name": "Api",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "ApprovalApiANYproxy80D0C580": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "ApprovalApi7AD56316",
        },
        "RouteKey": "ANY /{proxy+}",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "ApprovalApiANYproxyANYproxyIntegration9EB5C601",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "ApprovalApiANYproxyANYproxyIntegration9EB5C601": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "ApprovalApi7AD56316",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "ApprovalApprover72D44F63",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "ApprovalApiANYproxytestingstackApprovalApiANYproxy59DA5EE7Permission2A619A51": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "ApprovalApprover72D44F63",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "ApprovalApi7AD56316",
              },
              "/*/*/{proxy+}",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ApprovalApiDefaultStage1F3689E0": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "ApprovalApi7AD56316",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "ApprovalApprover72D44F63": Object {
      "DependsOn": Array [
        "ApprovalApproverServiceRole747D061A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters9df69b1c5244ca159360ab80f7e94d352de6fc13390ac375c21274ed8bed348dS3Bucket01FB5559",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters9df69b1c5244ca159360ab80f7e94d352de6fc13390ac375c21274ed8bed348dS3VersionKey4C30C2E2",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters9df69b1c5244ca159360ab80f7e94d352de6fc13390ac375c21274ed8bed348dS3VersionKey4C30C2E2",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_ACCOUNT_ID": Object {
              "Ref": "AWS::AccountId",
            },
            "SLACK_TOKEN": "xoxb-12345",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ApprovalApproverServiceRole747D061A",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ApprovalApproverServiceRole747D061A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "RequesterAllowInvoketestingstackTopic636EF16810BC6C79": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "RequesterC5D85B6B",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": Object {
          "Ref": "TopicBFC7AF6E",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RequesterC5D85B6B": Object {
      "DependsOn": Array [
        "RequesterServiceRole455136E7",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters9df69b1c5244ca159360ab80f7e94d352de6fc13390ac375c21274ed8bed348dS3Bucket01FB5559",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters9df69b1c5244ca159360ab80f7e94d352de6fc13390ac375c21274ed8bed348dS3VersionKey4C30C2E2",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters9df69b1c5244ca159360ab80f7e94d352de6fc13390ac375c21274ed8bed348dS3VersionKey4C30C2E2",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_ACCOUNT_ID": Object {
              "Ref": "AWS::AccountId",
            },
            "SLACK_TOKEN": "xoxb-12345",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "RequesterServiceRole455136E7",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "RequesterServiceRole455136E7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "RequesterTopic2A59E833": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "RequesterC5D85B6B",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": Object {
          "Ref": "TopicBFC7AF6E",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "TopicBFC7AF6E": Object {
      "Type": "AWS::SNS::Topic",
    },
  },
}
`;
